File: typescript/packages/a2a/src/protocol/pin.ts
import { sha256Hex } from "@threads/core/adapter";
import { AgentCard, schemeKind } from "./card";
import { fetchBytes, type Sending } from "./client";
import { speaks1_0 } from "./version";
import { type Binding, bindingOf, type Wire } from "./wire";

// Fetching a card and choosing what to speak, once. Whoever pins the result (a member's config
// bytes, or a thread's remote_card event) keeps the bytes, so a card that changes later never
// moves a conversation that has already started.

/** Why a remote cannot be used. Each is a value, never a throw, and each names the fix. */
export type PinFailure =
  | { readonly code: "remote_unavailable"; readonly message: string }
  | { readonly code: "remote_unsupported"; readonly message: string }
  | { readonly code: "remote_auth_unsupported"; readonly message: string };

export type PinnedCard = {
  readonly bytes: Uint8Array;
  readonly sha256: string;
  readonly card: AgentCard;
  readonly wire: Wire;
  /** `window_ms` from our idempotent-send extension, when the card declares it. */
  readonly dedupWindowMs: number | undefined;
  readonly streaming: boolean;
};

export const IDEMPOTENT_SEND =
  "https://threadsai.dev/a2a/ext/idempotent-send/v1";
export const PROVENANCE = "https://threadsai.dev/a2a/provenance/v1";

/** The bindings we speak, in the order we prefer them when a card offers both. */
const PREFERRED: readonly Binding[] = ["JSONRPC", "HTTP+JSON"];

/** A card over this many bytes is refused before it is parsed. */
export const MAX_CARD_BYTES: number = 256 * 1024;

export async function fetchCard(
  cardUrl: string,
  hasBearer: boolean,
  sending: Sending,
): Promise<PinnedCard | PinFailure> {
  const got = await fetchBytes(cardUrl, MAX_CARD_BYTES, sending);
  return got.kind === "failed"
    ? {
        code: "remote_unavailable",
        message: `${cardUrl} could not be read: ${got.why}`,
      }
    : pinCard(got.bytes, cardUrl, hasBearer);
}

/**
 * A card's exact bytes, parsed and checked. Kept separate from the fetch so materialize and a
 * rebind can verify pinned bytes without the network, which is what makes replay hermetic.
 */
export function pinCard(
  bytes: Uint8Array,
  cardUrl: string,
  hasBearer: boolean,
): PinnedCard | PinFailure {
  let json: unknown;
  try {
    json = JSON.parse(
      new TextDecoder(undefined, { fatal: true }).decode(bytes),
    );
  } catch {
    return {
      code: "remote_unavailable",
      message: `${cardUrl} is not a JSON agent card`,
    };
  }
  const parsed = AgentCard.safeParse(json);
  if (!parsed.success)
    return {
      code: "remote_unavailable",
      message: `${cardUrl} is not an A2A 1.0 agent card: ${parsed.error.message}`,
    };
  const card = parsed.data;
  const chosen = choose(card);
  if (chosen === undefined)
    return {
      code: "remote_unsupported",
      message: `${card.name} declares no 1.0 interface in a binding we speak (${offered(card)})`,
    };
  const auth = satisfiable(card, hasBearer);
  if (auth !== undefined) return auth;
  return {
    bytes,
    sha256: sha256Hex(bytes),
    card,
    wire: chosen,
    dedupWindowMs: windowOf(card),
    streaming: card.capabilities.streaming === true,
  };
}

/** The first interface that speaks 1.0 in a binding we speak, preferring JSON-RPC. */
function choose(card: AgentCard): Wire | undefined {
  for (const want of PREFERRED) {
    const found = card.supportedInterfaces.find(
      (i) =>
        speaks1_0(i.protocolVersion) && bindingOf(i.protocolBinding) === want,
    );
    if (found !== undefined) return { url: found.url, binding: want };
  }
  return undefined;
}

function offered(card: AgentCard): string {
  return card.supportedInterfaces
    .map((i) => `${i.protocolBinding} ${i.protocolVersion}`)
    .join(", ");
}

/**
 * Whether the credential we hold can satisfy the card. A card that declares no scheme asks for
 * nothing, so an unauthenticated call is what it wants. Otherwise one scheme must be an HTTP
 * `bearer`, because `bearer` is the only auth helper (decision 2).
 */
function satisfiable(
  card: AgentCard,
  hasBearer: boolean,
): PinFailure | undefined {
  const declared = Object.entries(card.securitySchemes ?? {});
  if (declared.length === 0) return undefined;
  const bearer = declared.some(
    ([, s]) =>
      schemeKind(s) === "httpAuthSecurityScheme" &&
      s.httpAuthSecurityScheme?.scheme.toLowerCase() === "bearer",
  );
  if (bearer && hasBearer) return undefined;
  const names = declared
    .map(([name, s]) => `${name} (${schemeKind(s) ?? "unrecognised"})`)
    .join(", ");
  return {
    code: "remote_auth_unsupported",
    message: bearer
      ? `${card.name} requires a bearer token; pass auth: bearer(secret("…"))`
      : `${card.name} declares only ${names}; bearer is the only scheme this adapter sends`,
  };
}

/** `window_ms` from our extension, when the card declares it with a usable value. */
function windowOf(card: AgentCard): number | undefined {
  const found = card.capabilities.extensions?.find(
    (e) => e.uri === IDEMPOTENT_SEND,
  );
  const raw = found?.params?.["window_ms"];
  return typeof raw === "number" && Number.isInteger(raw) && raw > 0
    ? raw
    : undefined;
}

File: python/src/threads/a2a/protocol/pin.py
"""Fetching a card and choosing what to speak, once. Whoever pins the result (a member's config
bytes, or a thread's `remote_card` event) keeps the bytes, so a card that changes later never
moves a conversation that has already started. `pin_card` is separate from `fetch_card` so pinned
bytes can be verified with no network, which is what makes replay hermetic."""

from dataclasses import dataclass
from typing import Final, Literal

from pydantic import ValidationError
from pydantic.experimental.missing_sentinel import MISSING

from threads._generated.a2a_v1 import AgentCard
from threads.a2a.protocol.card import is_bearer, scheme_kind
from threads.a2a.protocol.client import Sending, fetch_bytes
from threads.a2a.protocol.version import speaks_1_0
from threads.a2a.protocol.wire import Binding, Wire, binding_of
from threads.log.digest import sha256_hex

IDEMPOTENT_SEND: Final = "https://threadsai.dev/a2a/ext/idempotent-send/v1"
PROVENANCE: Final = "https://threadsai.dev/a2a/provenance/v1"

MAX_CARD_BYTES: Final = 256 * 1024
"""A card over this many bytes is refused before it is parsed."""

_PREFERRED: Final[tuple[Binding, ...]] = ("JSONRPC", "HTTP+JSON")
"""The bindings we speak, in the order we prefer them when a card offers both."""

type PinFailureCode = Literal["remote_unavailable", "remote_unsupported", "remote_auth_unsupported"]


@dataclass(frozen=True, slots=True)
class PinFailure:
    """Why a remote cannot be used. A value, never a raise, and each message names the fix."""

    code: PinFailureCode
    message: str


@dataclass(frozen=True, slots=True)
class PinnedCard:
    bytes_: bytes
    sha256: str
    card: AgentCard
    wire: Wire
    dedup_window_ms: int | None
    """`window_ms` from our idempotent-send extension, when the card declares it."""
    streaming: bool


async def fetch_card(card_url: str, has_bearer: bool, sending: Sending) -> PinnedCard | PinFailure:
    got = await fetch_bytes(card_url, MAX_CARD_BYTES, sending)
    if got.bytes_ is None:
        return PinFailure("remote_unavailable", f"{card_url} could not be read: {got.why}")
    return pin_card(got.bytes_, card_url, has_bearer)


def pin_card(raw: bytes, card_url: str, has_bearer: bool) -> PinnedCard | PinFailure:
    """A card's exact bytes, parsed and checked."""
    try:
        card = AgentCard.model_validate_json(raw)
    except (ValidationError, ValueError) as error:
        return PinFailure("remote_unavailable", f"{card_url} is not an A2A 1.0 agent card: {error}")
    chosen = _choose(card)
    if chosen is None:
        return PinFailure(
            "remote_unsupported",
            f"{card.name} declares no 1.0 interface in a binding we speak ({_offered(card)})",
        )
    refused = _satisfiable(card, has_bearer=has_bearer)
    if refused is not None:
        return refused
    return PinnedCard(
        raw,
        sha256_hex(raw),
        card,
        chosen,
        _window_of(card),
        card.capabilities.streaming is True,
    )


def _choose(card: AgentCard) -> Wire | None:
    """The first interface that speaks 1.0 in a binding we speak, preferring JSON-RPC."""
    for want in _PREFERRED:
        for offered in card.supportedInterfaces:
            if speaks_1_0(offered.protocolVersion) and binding_of(offered.protocolBinding) == want:
                return Wire(offered.url, want)
    return None


def _offered(card: AgentCard) -> str:
    return ", ".join(f"{i.protocolBinding} {i.protocolVersion}" for i in card.supportedInterfaces)


def _satisfiable(card: AgentCard, *, has_bearer: bool) -> PinFailure | None:
    """Whether the credential we hold can satisfy the card. A card that declares no scheme asks for
    nothing, so an unauthenticated call is what it wants. Otherwise one scheme must be an HTTP
    `bearer`, because `bearer` is the only auth helper."""
    declared = {} if card.securitySchemes is MISSING else dict(card.securitySchemes)
    if not declared:
        return None
    bearer = any(is_bearer(scheme) for scheme in declared.values())
    if bearer and has_bearer:
        return None
    if bearer:
        return PinFailure(
            "remote_auth_unsupported",
            f'{card.name} requires a bearer token; pass auth=bearer(secret("..."))',
        )
    names = ", ".join(
        f"{name} ({scheme_kind(scheme) or 'unrecognised'})" for name, scheme in declared.items()
    )
    return PinFailure(
        "remote_auth_unsupported",
        f"{card.name} declares only {names}; bearer is the only scheme this adapter sends",
    )


def _window_of(card: AgentCard) -> int | None:
    """`window_ms` from our extension, when the card declares it with a usable value."""
    extensions = () if card.capabilities.extensions is MISSING else card.capabilities.extensions
    for extension in extensions:
        if extension.uri != IDEMPOTENT_SEND or extension.params is MISSING:
            continue
        raw = extension.params.get("window_ms")
        if isinstance(raw, int) and not isinstance(raw, bool) and raw > 0:
            return raw
    return None
